### PR TITLE
Add IPv6 BIG TCP support

### PR DESCRIPTION
**Pull request by NikAleksandrov**
Add a new IPv6 BIG TCP[1] option and infrastructure which allows to change a device's GRO and GSO max sizes depending on their current values and the new option (enable-ipv6-big-tcp). BIG TCP is used to allow the network stack to cook bigger TSO/GRO packets (512k is the new maximum, 64k was the old), it can be used only with IPv6 and improves latency and performance due to the reduced number of packets that traverse the stack. It works by adding a new temporary Hop by Hop header right after the IPv6 header, it is stripped before the packet is sent on the wire.

The defaults (65536) are set only if necessary (i.e. if the GSO/GRO max sizes were previously changed), when the option is enabled they get set to 196608. The bpf_dynptr_data helper is used to probe for 5.19+ kernel. BIG TCP can be used only in BPF host routing mode, without encryption and tunneling, there are checks that enforce compatibility. If BIG TCP cannot tune all external interfaces then it will revert back to the defaults. Currently it is supported only on mlx4/mlx5 and veth devices, also any device that inherits gro/gso maximum sizes from others (e.g. bonding should work).

LWN also has a nice BIG TCP summary available [here](https://lwn.net/Articles/884104/)

If the new BIG TCP Cilium option is enabled and initialized it can be seen in the logs:
level=info msg="Setting up IPv6 BIG TCP" subsys=big-tcp
level=info msg="Setting gso_max_size to 196608 and gro_max_size to 196608" device=eth0 subsys=big-tcp

Netperf benchmarks:

```
- TCP_RR
MIN_LATENCY  P90_LATENCY  P99_LATENCY  THROUGHPUT (trans/sec)

BIG TCP Disabled
60            91           136         13087.53

BIG-TCP Enabled
38            76           109         15629.73

- TCP_STREAM (Mbps)
16386.21 (BIG TCP disabled)
23181.41 (BIG TCP enabled)
```

Standalone tests which use a Kind cluster are added. BIG TCP currently only works with veth and mellanox devices
so we need a Kind cluster to test it properly, I've left a TODO note that it should be migrated to the new infra ones it
lands.

Note that pods need to  be restarted when the option is changed due to the veth configuration which gets applied when the
devices are created and before they're put in the target netns.

Patch-set overview:
 Patch 01 - adds basic BIG TCP infrastructure allowing to enable/disable it and check for requirements
 Patch 02 - adds GRO/GSO max sizes to the daemon configuration status so they can be exposed for veth device configuration
 Patch 03 - adds the new "--enable-ipv6-big-tcp" option to the daemon
 Patch 04 - uses the exposed GRO/GSO max sizes to configure the newly created pod veth devices
 Patch 05 - adds helm support for the new option (enableIPv6BIGTCP)
 Patch 06 - exposes the IPv6 BIG TCP state in "cilium status" (IPv6 BIG TCP:            Enabled/Disabled)
 Patch 07 - documents the new feature in the tuning docs
 Patch 08 - adds BIG TCP tests which provision a Kind cluster, verify the option is properly enabled and run netperf

Requirements to enable:
* Kernel >= 5.19
* eBPF Host-Routing
* eBPF-based kube-proxy replacement
* eBPF-based masquerading
* Tunneling and encryption disabled
* Supported NICs: mlx4, mlx5

Thanks,
  Nik

[1] https://lore.kernel.org/netdev/20220513183408.686447-3-eric.dumazet@gmail.com/T/

### CODE REVIEW

**Comment by NikAleksandrov** [comment]
I accidentally pushed the wrong auto-generated model file, sorry about that.
I've pushed the correct one now.

File: install/kubernetes/cilium/values.yaml [hunk 1]
```diff
@@ -206,7 +206,6 @@ azure:
   # Note that this is incompatible with AKS clusters created in BYOCNI mode: use
   # AKS BYOCNI integration (`aksbyocni.enabled`) instead.
   enabled: false
-  # usePrimaryAddress: false
```
Comment (borkmann) [inline install/kubernetes/cilium/values.yaml]: unrelated?

**Review by borkmann — APPROVED** [review]
LGTM! Great work 🚀

File: test/bigtcp/test.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
```
Comment (brb) [inline test/bigtcp/test.sh:1]: You might want to run this test as part of the CI L4LB. To make it happen, you need to add exec the test.sh here: https://github.com/cilium/cilium/blob/master/.github/workflows/tests-l4lb.yaml#L198

**Review by brb — CHANGES_REQUESTED** [review]
Nice PR! Reviewed the testing bits.

File: test/bigtcp/test.sh (hunk 2 above)
Comment (NikAleksandrov) [reply test/bigtcp/test.sh:1]: ah nice catch, I'll update the test commit to include it, thanks!

File: install/kubernetes/cilium/values.yaml (hunk 1 above)
Comment (NikAleksandrov) [reply install/kubernetes/cilium/values.yaml]: it was because of "make -C install/kubernetes cilium/values.yaml"
one of the tests was failing without it

Comment (NikAleksandrov) [reply install/kubernetes/cilium/values.yaml]: anyway, it seems after rebasing on master that isn't needed anymore so I've dropped it

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		log.WithError(err).WithField("device", device).Warn("Link does not exist")
+		return nil
+	}
+
+	attrs := link.Attrs()
+	if (int(attrs.GROMaxSize) == GROMaxSize && int(attrs.GSOMaxSize) == GSOMaxSize) ||
+		(!option.Config.EnableIPv6BIGTCP &&
+			int(attrs.GROMaxSize) <= GROMaxSize &&
+			int(attrs.GSOMaxSize) <= GSOMaxSize) {
+		return nil
```
Comment (jrfastab) [inline pkg/datapath/linux/bigtcp/bigtcp.go:67]: The second case here, 
```
!EnableIPv6BIGTCP && attrs.GROMaxSize <= GROMaxSize && GSOMaxSize <= GSOMaxSize
```
this is an error right? The user tried to set GRO/GSO larger than the max should we return an error?

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	err = netlink.LinkSetGROMaxSize(link, GROMaxSize)
+	if err != nil {
+		return err
+	}
+
+	return netlink.LinkSetGSOMaxSize(link, GSOMaxSize)
```
Comment (jrfastab) [inline pkg/datapath/linux/bigtcp/bigtcp.go:75]: If this errors should we unwinde the GROMaxSize case so this routine doesn't leave us in some partial state.

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		return
+	}
+
+	kernelGood := false
+	if h := probes.NewProbeManager().GetHelpers("sched_cls"); h != nil {
+		_, kernelGood = h["bpf_dynptr_data"]
+	}
+	if !kernelGood {
+		if option.Config.EnableIPv6BIGTCP {
+			log.Fatalf("Cannot enable --%s, needs kernel 5.19 or newer",
```
Comment (jrfastab) [inline pkg/datapath/linux/bigtcp/bigtcp.go]: Do we really want to fatalf here? This seems extreme to me. In alpo for example we have a mix of kernel versions and we might upgrade some nodes to 5.19 when its available but this will stop us from enabling it in configMap. I would suggest if we really want to fatalf here we should have 'strict' and 'probe' modes.

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		}
+		if option.Config.TunnelingEnabled() {
+			log.Fatal("IPv6 BIG TCP is not supported in tunneling mode")
+		}
+		if option.Config.EncryptionEnabled() {
+			log.Fatal("IPv6 BIG TCP is not supported with encryption enabled")
+		}
+		if option.Config.EnableHostLegacyRouting {
+			log.Fatal("IPv6 BIG TCP is not supported with legacy host routing")
+		}
```
Comment (jrfastab) [inline pkg/datapath/linux/bigtcp/bigtcp.go:121]: OTOH these fatal events seem OK to me because this is a config error that needs to be resolved.

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 4 above)
Comment (jrfastab) [reply pkg/datapath/linux/bigtcp/bigtcp.go:75]: ok I see you roll back in the caller now. I guess this is OK. Maybe add a comment to the function so next dev that comes along knows they need to do rollback from caller as well.

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+			option.Config.EnableIPv6BIGTCP = false
+			break
+		}
+		log.WithField("device", device).Infof("Setting gso_max_size to %d and gro_max_size to %d",
+			bigtcpConfig.gsoMaxSize, bigtcpConfig.groMaxSize)
+	}
+
+	if err != nil {
+		bigtcpConfig.groMaxSize = defaultGROMaxSize
+		bigtcpConfig.gsoMaxSize = defaultGSOMaxSize
```
Comment (jrfastab) [inline pkg/datapath/linux/bigtcp/bigtcp.go]: Looks like default*Size is just a var  const. Probably to be save should read these out at top of function so if something else has configured values we don't whack them here on error. GKE for example does some /proc settings I would guess others do as well.

File: daemon/cmd/daemon_main.go [hunk 8]
```diff
@@ -725,6 +725,9 @@ func initializeFlags() {
 	flags.Bool(option.EnableIPMasqAgent, false, "Enable BPF ip-masq-agent")
 	option.BindEnv(option.EnableIPMasqAgent)
 
+	flags.Bool(option.EnableIPv6BIGTCP, false, "Enable IPv6 BIG TCP option which increases device's maximum GRO/GSO limits")
```
Comment (jrfastab) [inline daemon/cmd/daemon_main.go:727]: maybe make a tri-state to handle case where we fatal fail if we can't configure this?

**Comment by jrfastab** [comment]
Another thought, probing for kernel version through bpf helper is a bit clumsy. Can't we just do a GetKernelVersion() routine and check it directly. We have the code in Tetragon to do the kernel version check if you want to rip it out.

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 3 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go:67]: no, that's to avoid changing gso/gro when not necessary, we have the default case where big tcp is disabled and the case where it was enabled and then disabled, this is used to match the first one

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 4 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go:75]: sure, will do

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 5 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: we can just disable the option with a log msg, but people must look for it to verify what happened, I think a hard error is better so they can't miss it, that's why it fatalfs in the requirement checks

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 6 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go:121]: ah right, that's what I meant

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 7 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: the problem is once it's enabled we don't know what it was before that, we could record them but it seems like an overkill, gso and gro max sizes are rarely changed from their 64k defaults,
I'll think about it more and will check if any of the vendors changes them

edit: ahh I missed the error part, yeah we could do that for the error path. TBH I very much doubt anyone is reducing gro or gso max sizes. :)

File: daemon/cmd/daemon_main.go (hunk 8 above)
Comment (NikAleksandrov) [reply daemon/cmd/daemon_main.go:727]: it already does that in most cases when requirements can't be met, am I missing some case or did I misunderstand you?

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 3 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go:67]: I'll add a comment

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 4 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go:75]: added a comment to make sure devs know about rolling back any changes that have been done

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 7 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: By the way note that these settings are per-device and can only be changed via netlink calls (they aren't exported to /proc or /sys)

Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: I left a comment in big tcp's docs about the assumptions we make and the logic of enable/disable (when Cilium actually does any value changes). I think that should be ok for the time being, if we hit any issues we can revisit the logic and perhaps persist the old values to be sure that we need to change them to anything different.

**Comment by borkmann** [comment]
/test-1.16-4.9

**Review by brb — APPROVED** [review]
The CI changes LGTM. Thanks.

File: cilium-health/launch/endpoint.go [hunk 9]
```diff
@@ -272,7 +272,7 @@ func LaunchAsEndpoint(baseCtx context.Context,
 
 	switch option.Config.DatapathMode {
 	case datapathOption.DatapathModeVeth:
-		_, epLink, err := connector.SetupVethWithNames(vethName, epIfaceName, mtuConfig.GetDeviceMTU(), info)
+		_, epLink, err := connector.SetupVethWithNames(vethName, epIfaceName, mtuConfig.GetDeviceMTU(), 0, 0, info)
```
Comment (gandro) [inline cilium-health/launch/endpoint.go]: Any particular reason as to why you are not passing in `d.bigtcpConfig` here? `LaunchAsEndpoint` is called from the Daemon, so it should be fairly straight forward to get it.

Now I understand that the health does not send big packets, so there is absolutely no performance benefit to do it. But on the other hand, the health endpoint should imitate the regular pod setup as much as possible, so we catch connectivity issues there early. I'm fine to keep the code as is, but I am curious if you have given this any thought.

**Review by gandro — CHANGES_REQUESTED** [review]
Reviewed Helm and Health. Both look good to me, one high-level question around health however.

(edit: I misclicked. I meant to approve, no blocking changes requested)

**Review by gandro — APPROVED** [review]

File: cilium-health/launch/endpoint.go (hunk 9 above)
Comment (NikAleksandrov) [reply cilium-health/launch/endpoint.go]: IIRC I didn't want to add any additional unnecessary direct dependencies on datapath/linux/bigtcp and to lesser extent also to be consistent with the rest of the code which takes two integers for GSO and GRO.

Comment (gandro) [reply cilium-health/launch/endpoint.go]: Not sure I follow. You can decouple the package dependency by adding an interface for `Get{GRO,GSO}MaxSize`, similar to e.g. the `routingConfigurer` interface. Or you could just pass in two integers if you find that more consistent?

Comment (NikAleksandrov) [reply cilium-health/launch/endpoint.go]: I don't think the health monitor needs large gso/gro limits as you've also noted, keeping it as-is seems marginally simpler, but I don't have a strong preference and having config parity sounds good so I  don't mind passing the real values. I'll push the change after running a few tests locally.

Comment (NikAleksandrov) [reply cilium-health/launch/endpoint.go]: I've pushed the changes in this commit which also adds the support for configuring gso/gro limits on a veth pair.
Thanks for the review!

File: Documentation/operations/performance/tuning.rst [hunk 10]
```diff
@@ -37,6 +37,55 @@ in any of the Cilium pods and look for the line reporting the status for
 * eBPF-based kube-proxy replacement
 * eBPF-based masquerading
 
+IPv6 BIG TCP
+============
+
+IPv6 BIG TCP allows the network stack to prepare larger TSO (transmit) and GRO
```
Comment (qmonnet) [inline Documentation/operations/performance/tuning.rst]: Nit: s/TSO/GSO/ ? I know “TSO” exists, but the rest of the document only mentions GSO (and conversely, it uses GSO later without mentioning it's for transmitting).

File: Documentation/operations/performance/tuning.rst [hunk 11]
```diff
@@ -37,6 +37,55 @@ in any of the Cilium pods and look for the line reporting the status for
...
+setup, IPv4 packets will use the old lower limits (64k) and IPv6 packets will
+use the new larger ones (192k). Note that Cilium assumes the default kernel values
+for GSO and GRO maximum sizes are 64k and adjusts them only when necessary, i.e. if
+BIG TCP is enabled and the current GSO/GRO maximum sizes are less than 192k it
+will try to increase them, respectively when BIG TCP is disabled and the current
+maximum values are more than 64k it will try to decrease them.
+
+**Requirements:**
+
+* Kernel >= 5.19
```
Comment (qmonnet) [inline Documentation/operations/performance/tuning.rst:59]: Maybe worth adding a line on kernel requirements in https://docs.cilium.io/en/latest/operations/system_requirements/#required-kernel-versions-for-advanced-features as well?

File: test/bigtcp/test.sh (hunk 2 above)
Comment (qmonnet) [inline test/bigtcp/test.sh:1]: Please add the license/copyright header. I know not all bash scripts have it, but let's try to add it for new files.

**Review by qmonnet — APPROVED** [review]
The code looks good. Nice work, thanks! PR is well split and easy to review.

I've got some nitpicks on the doc (non-blocking), please see below.

File: cilium-health/launch/endpoint.go (hunk 9 above)
Comment (gandro) [reply cilium-health/launch/endpoint.go]: Awesome, thanks Nikolay!

File: test/bigtcp/test.sh (hunk 2 above)
Comment (NikAleksandrov) [reply test/bigtcp/test.sh:1]: ack, will do

File: Documentation/operations/performance/tuning.rst (hunk 11 above)
Comment (NikAleksandrov) [reply Documentation/operations/performance/tuning.rst:59]: sounds good, I'll add that as well

File: Documentation/operations/performance/tuning.rst (hunk 10 above)
Comment (NikAleksandrov) [reply Documentation/operations/performance/tuning.rst]: That was intentional actually because GSO segments are later passed down for TSO, but I agree it's best to be consistent.
I'll change it, thanks!

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 5 above)
Comment (jrfastab) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: it will break deployment into our dogfooding cluster though.

Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: ok, I've pushed a change to make it a warning with a message why it couldn't be enabled if the kernel is incompatible

File: Documentation/cmdref/cilium-agent.md [hunk 12]
```diff
@@ -107,6 +107,7 @@ cilium-agent [flags]
       --enable-ipv4-fragment-tracking                           Enable IPv4 fragments tracking for L4-based lookups (default true)
       --enable-ipv4-masquerade                                  Masquerade IPv4 traffic from endpoints leaving the host (default true)
       --enable-ipv6                                             Enable IPv6 support (default true)
+      --enable-ipv6-big-tcp                                     Enable IPv6 BIG TCP option which increases device's maximum GRO/GSO limits
```
Comment (tommyp1ckles) [inline Documentation/cmdref/cilium-agent.md:110]:
```suggestion
      --enable-ipv6-big-tcp                                     Enable IPv6 BIG TCP option which increases device's maximum GRO/GSO limits (default false)
```

File: pkg/client/client.go [hunk 13]
```diff
@@ -381,6 +381,14 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
 		}
 	}
 
+	if sr.IPV6BigTCP != nil {
+		status := "Enabled"
+		if !sr.IPV6BigTCP.Enabled {
+			status = "Disabled"
+		}
+		fmt.Fprintf(w, "IPv6 BIG TCP:\t%s\n", status)
```
Comment (tommyp1ckles) [inline pkg/client/client.go:389]:
```suggestion
		fmt.Fprintf(w, "IPv6 Big TCP:\t%s\n", status)
```

File: pkg/datapath/connector/veth.go [hunk 14]
```diff
@@ -123,6 +123,28 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
 		return nil, nil, fmt.Errorf("unable to bring up veth pair: %s", err)
 	}
 
+	if groMaxSize > 0 {
+		if err = netlink.LinkSetGROMaxSize(hostVeth, groMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GRO max size to %q: %s",
```
Comment (tommyp1ckles) [inline pkg/datapath/connector/veth.go]: I would use error wrapping for these, same goes for the 3 returns with fmt.Errorf below.
```suggestion
			return nil, nil, fmt.Errorf("unable to set GRO max size to %q: %w",
```

File: pkg/datapath/connector/veth.go [hunk 15]
```diff
@@ -123,6 +123,28 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
...
+				lxcIfName, err)
+		}
+		if err = netlink.LinkSetGROMaxSize(peer, groMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GRO max size to %q: %s",
```
Comment (tommyp1ckles) [inline pkg/datapath/connector/veth.go]: Should we distinguish between the host & peer link in either of these error cases messages?

File: pkg/datapath/connector/veth.go [hunk 16]
```diff
@@ -123,6 +123,28 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
...
+		}
+	}
+
+	if gsoMaxSize > 0 {
+		if err = netlink.LinkSetGSOMaxSize(hostVeth, gsoMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GSO max size to %q: %s",
+				lxcIfName, err)
+		}
+		if err = netlink.LinkSetGSOMaxSize(peer, gsoMaxSize); err != nil {
+			return nil, nil, fmt.Errorf("unable to set GSO max size to %q: %s",
```
Comment (tommyp1ckles) [inline pkg/datapath/connector/veth.go]: Same question as above about distinguishing between links.

File: pkg/datapath/connector/veth.go (hunk 15 above)
Comment (NikAleksandrov) [reply pkg/datapath/connector/veth.go]: No, we don't care which one failed. This is handled the same for both by the driver, if one fails then the other will fail as well.

File: Documentation/cmdref/cilium-agent.md (hunk 12 above)
Comment (NikAleksandrov) [reply Documentation/cmdref/cilium-agent.md:110]: ack, I've added the default

File: pkg/datapath/connector/veth.go (hunk 14 above)
Comment (NikAleksandrov) [reply pkg/datapath/connector/veth.go]: ack, changed

File: Documentation/cmdref/cilium-agent.md (hunk 12 above)
Comment (NikAleksandrov) [reply Documentation/cmdref/cilium-agent.md:110]: actually no, the default was there in the beginning and when building the documentation checks remove it, if I add it back the docs test test below fails so I've removed it again

**Review by tommyp1ckles — APPROVED** [review]

File: api/v1/openapi.yaml [hunk 17]
```diff
@@ -2821,6 +2830,16 @@ definitions:
       forwarded:
         description: Number of messages forwarded
         type: integer
+  IPV6BigTCP:
```
Comment (tklauser) [inline api/v1/openapi.yaml:2833]: Nit:

```suggestion
  IPv6BigTCP:
```

File: daemon/cmd/daemon.go [hunk 18]
```diff
@@ -195,6 +196,9 @@ type Daemon struct {
 
 	// Controllers owned by the daemon
 	controllers *controller.Manager
+
+	// BIG-TCP config values
+	bigtcpConfig bigtcp.Configuration
```
Comment (tklauser) [inline daemon/cmd/daemon.go]:
```suggestion
	bigTCPConfig bigtcp.Configuration
```

File: pkg/datapath/linux/bigtcp/bigtcp.go [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+const (
+	defaultGROMaxSize int = 65536
+	defaultGSOMaxSize int = 65536
+
+	BIGTCPGROMaxSize int = 196608
+	BIGTCPGSOMaxSize int = 196608
```
Comment (tklauser) [inline pkg/datapath/linux/bigtcp/bigtcp.go]: These do not seem to be used outside the package, so:

```suggestion
	bigTCPGROMaxSize int = 196608
	bigTCPGSOMaxSize int = 196608
```

Comment (tklauser) [inline pkg/datapath/linux/bigtcp/bigtcp.go]: It's more common in Go to declare numeric constants as untyped. That way they can be used with arbitrary integer types depending on context. Any particular reason these need to be `int`s? I think making them untyped would allow to get rid of the type conversions in `setGROGSOMaxSize` below.

```suggestion
	defaultGROMaxSize = 65536
	defaultGSOMaxSize = 65536

	BIGTCPGROMaxSize = 196608
	BIGTCPGSOMaxSize = 196608
```

File: api/v1/openapi.yaml (hunk 17 above)
Comment (NikAleksandrov) [reply api/v1/openapi.yaml:2833]: this was changed automatically by something that checks the openapi file, I had it as IPv6BigTCP and it failed a test here

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 19 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: no particular reason or preference, I'll change them

File: daemon/cmd/daemon.go (hunk 18 above)
Comment (NikAleksandrov) [reply daemon/cmd/daemon.go]: sure, why not :)

File: pkg/datapath/linux/bigtcp/bigtcp.go (hunk 19 above)
Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: They weren't big for export purposes (although that happens), but because that's the real name of the feature BIG-TCP.
Anyway, I'll make them small to avoid exporting them

Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: Actually they aren't used in setGROGSOMaxSize() directly, maxGRO/GSO are passed as integer arguments
because the netlink library takes GSO/GRO values as ints.

Comment (tklauser) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: Right, the type conversions in setGROGSOMaxSize are still needed then. But I think the consts can remain untyped.

Comment (NikAleksandrov) [reply pkg/datapath/linux/bigtcp/bigtcp.go]: Yep, they can. I'll push the changes in a minute

**Review by rolinh — APPROVED** [review]
LGTM for the vendor and API changes :+1:

**Review by tklauser — APPROVED** [review]

**Comment by NikAleksandrov** [comment]
the ConformanceAKS test failure is unrelated to BIG TCP. It fails at Cilium Cleanup, details:
```
Run pkill -f "cilium.*hubble.*port-forward|kubectl.*port-forward.*hubble-relay"
Error: Process completed with exit code 1.
```
